### PR TITLE
backend/coins/eth: add fee targets based on EthGasStation

### PR DESCRIPTION
**Pull request by benma**
https://ethgasstation.info/ provides gas price estimates for different
priorities.

We use these to give the user fee options.

If the service is not available, we fall back to the previous gas
price estimation provided by the eth_gasPrice RPC call to the ETH
node (currently etherscan proxy).

FeeTargets() is currently doing all the work whenever it is called,
which should be once when the send screen is loaded and every time a
tx proposal is made (every time the user changes something in the send
screen). This could be optimized with a cache that is invalidated
regularly (or when a new block arrives).

Custom gas price option to be added later.

### CODE REVIEW

File: backend/coins/eth/feetarget.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+// FormattedFeeRate returns a string showing the fee rate.
+func (f *feeTarget) FormattedFeeRate() string {
+	if f.gasPrice == nil {
+		return ""
+	}
+	factor := big.NewInt(1e9)
+	return strings.TrimRight(strings.TrimRight(
+		new(big.Rat).SetFrac(f.gasPrice, factor).FloatString(9),
+		"0"), ".") + " Gwei"
```
Comment (x1ddos) [inline backend/coins/eth/feetarget.go:46]: spent at least few minutes trying to match parenthesis. is the idea to remove trailing zeros? this should be equivalent but easier to read imho:

```go
s := new(big.Rat).SetFrac(f.gasPrice, factor).FloatString(9)
return strings.TrimRight(s, ".0") + " Gwei"
```

File: frontends/web/src/locales/en/app.json [hunk 2]
```diff
@@ -1134,12 +1135,16 @@
     "feeTarget": {
       "description": {
         "economy": "4 hours (24 blocks)",
+        "economy_eth": "< 30 minutes",
```
Comment (x1ddos) [inline frontends/web/src/locales/en/app.json:1138]: might be misleading. I would spell it out: "30 minutes or less"

File: backend/coins/eth/account.go [hunk 3]
```diff
@@ -427,34 +434,30 @@ type TxProposal struct {
 	Keypath signing.AbsoluteKeypath
 }
 
-func (account *Account) newTx(
-	recipientAddress string,
-	amount coin.SendAmount,
-	data []byte,
-) (*TxProposal, error) {
-	if !ethcommon.IsHexAddress(recipientAddress) {
+func (account *Account) newTx(args *accounts.TxProposalArgs) (*TxProposal, error) {
```
Comment (x1ddos) [inline backend/coins/eth/account.go:437]: TxProposalArgs doc comment still says:

> // Only applies to BTC/LTC.

**Review by x1ddos — APPROVED** [review]
LGTM. Yeah, looks pretty much straightforward.

File: backend/coins/eth/account.go (hunk 3 above)
Comment (x1ddos) [reply backend/coins/eth/account.go:437]: ehm, I meant the `FeeTargetCode` field.

File: backend/coins/eth/feetarget.go (hunk 1 above)
Comment (benma) [reply backend/coins/eth/feetarget.go:46]: Your code turns `10` into `1`.

We use this type of formatting in a few places in the app and also in the firmware: https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/rust/util/src/decimal.rs#L32

I'll try to add unit tests.

Comment (benma) [reply backend/coins/eth/feetarget.go:46]: Ah it doesn't cause FloatString adds decimal places. Hmm

**Comment by thisconnect** [comment]
Tested on testnet and got different fee options, looks good  👍

File: backend/coins/eth/feetarget.go (hunk 1 above)
Comment (x1ddos) [reply backend/coins/eth/feetarget.go:46]: > Your code turns 10 into 1.

sounds like even more reason to comment the code and make it more readable.

Comment (benma) [reply backend/coins/eth/feetarget.go:46]: I added unit tests, your code doesn't work for e.g. `21e10` which becomes `210 Gwei` - source code strips beyond the decimal place to `21 Gwei`.

File: frontends/web/src/locales/en/app.json (hunk 2 above)
Comment (benma) [reply frontends/web/src/locales/en/app.json]: done

File: backend/coins/eth/account.go (hunk 3 above)
Comment (benma) [reply backend/coins/eth/account.go:439]: removed

**Comment by benma** [comment]
> Tested on testnet and got different fee options, looks good +1

Hmm I forgot about testnet and that is unexpected - I call the same endpoints. I'll make sure it only happens on mainnet. Good catch!

File: backend/coins/eth/account.go [hunk 4]
```diff
@@ -44,6 +46,8 @@ import (
 
 var pollInterval = 60 * time.Second
 
+const ethGasStationAPIKey = "289ea7884cb5903dcb50af9a81cd9518ae000ff9088af5182b5bc8b87385"
```
[the remark this answers is not in the document]
Comment (benma) [reply backend/coins/eth/account.go:51]: commented

File: backend/coins/eth/feetarget.go (hunk 1 above)
Comment (x1ddos) [reply backend/coins/eth/feetarget.go:46]: those 3 lines are still unreadable.

Comment (benma) [reply backend/coins/eth/feetarget.go]: improved

Comment (x1ddos) [reply backend/coins/eth/feetarget.go]: much nicer. thanks! :+1:

**Comment by benma** [comment]
> > Tested on testnet and got different fee options, looks good +1
> 
> Hmm I forgot about testnet and that is unexpected - I call the same endpoints. I'll make sure it only happens on mainnet. Good catch!

On 2nd thought, I'll just leave it. Less complicated, and it doesn't matter what the fees are on testnet really.

**Comment by thisconnect** [comment]
> On 2nd thought, I'll just leave it. Less complicated, and it doesn't matter what the fees are on testnet really.

👍  I prefer it too, we could also consider using btc mainnet fees on the testnet but just a thought... unless there is a good reason to not do it.

**Comment by x1ddos** [comment]
Forgot to mention: update changelog.

**Comment by benma** [comment]
> Forgot to mention: update changelog.

Great catch. Added in https://github.com/digitalbitbox/bitbox-wallet-app/pull/1410/

**Pull request merged**